### PR TITLE
Bugfix/programming exercise/fix broken endpoints

### PR DESCRIPTION
**Pull request by thilo-behnke**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
:warning: This branch does not run on the test server atm, even though the production build works locally.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] ~Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)~
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] ~Client: I added multiple integration tests (Jest) related to the features~
- [x] ~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~
- [x] ~Client: I documented the TypeScript code using JSDoc style.~
- [x] ~Client: I added multiple screenshots/screencasts of my UI changes~
- [x] ~Client: I translated all the newly inserted strings into German and English~

### Description

Fix issue from: https://github.com/ls1intum/Artemis/issues/837

- [x] 1) Bugfix updateProblemStatement (e.g. in edit in editor)
- [x] 2) Bugfix deleteProgrammingExercise (Course Administration > Delete)
- [x] 3) Hotfix updateProgrammingExercise (Course Administration > Edit > Save): Issue with jhipster-dependency-update: https://github.com/jhipster/jhipster-dependencies/blob/v0.x_maintenance/pom.xml, rollback is the only quick fix I found. https://github.com/ls1intum/Artemis/pull/838/files#diff-503f218d646c10f484fdc9d6315bf2e3R10
- [x] Help needed: Is this a good fix for the 3rd issue? Why does the current version of the branch not work on the test server? => solved, thx to @krusche 

### Steps for Testing

**Edit ProgrammingExercise:**
1. Log in to Artemis
2. Navigate to Course Administration > Programming Exercises > Edit
3. Click save at the bottom => should update the entity

**Edit ProgrammingExercise ProblemStatement:**
1. Log in to Artemis
2. Navigate to Course Administration > Programming Exercises > Edit > Update problem statement
3. Click save at the top of the editor => should update the problem statement

**Delete ProgrammingExercise**
1. Log in to Artemis
2. Navigate to Course Administration > Programming Exercises > Delete
3. Confirm delete => Programming Exercise should be removed from db

### CODE REVIEW

**Comment by thilo-behnke** [comment]
**Current state of the PR:**
All 3 issues are fixed, but the app does not start on the test server (prod build works locally).
Also to fix the 3rd issue I had to rollback a dep update from jhipster.
I didn't assign any other reviewers because we probably have to discuss how to approach this issue.

**Comment by krusche** [comment]
I would say #3 is not a good solution: Which dependency exactly is the problem here?

**Comment by krusche** [comment]
I just looked into `ProgrammingExerciseResource.updateProgrammingExercise(...)` and to be honest, I think the implementation is quite bad. I think we should refactor it:
* I don't think we even need to update/save the templateParticipation and solutionParticipation. During edit, they **cannot** change.
* Why do we even handle `Exercise.participations`? I also do think that nothing needs to be done here, because participations **cannot** change while editing a programming exercise

I will try out what happens if I remove this code.

**Comment by krusche** [comment]
Thanks for identifying these issues!

I was able to fix problem 3) in `develop`. Deleting programming exercises also works fine for me, not sure what the exact problem was here. I also tested creating programming exercises which works fine, but I did not test all other areas for programming exercises in the course administration.

I will now update this branch with the changes from `develop`. It might be necessary to adapt tests. 

@thilo-behnke could you please revert the changes regarding the dependencies and check if your problem 1) is solved properly? Thank you!

I will deploy a hotfix release 3.7.1 later today to production.

**Comment by krusche** [comment]
I just tested it on the test server and updating the problem statement works for me fine, so I guess the problem is solved and we could even close this PR.

The only issue now is that suddenly many Java tests fail due to some database integrity reasons. Would be great if you could have a look

**Comment by krusche** [comment]
I was able to fix the tests. There was a simple problem when cleaning up the database.

**Comment by krusche** [comment]
I guess the new tests in this PR and maybe some code refactoring are worth to be merge to `develop`, so I keep this PR open.

**Comment by thilo-behnke** [comment]
@krusche Thanks for looking into the issue and solving it! I'll prepare this branch now for the merge.

File: src/test/java/de/tum/in/www1/artemis/service/ProgrammingExerciseScheduleServiceTest.java [hunk 1]
```diff
@@ -47,7 +47,7 @@
 
     // When the scheduler is invoked, there is a small delay until the runnable is called.
     // TODO: This could be improved by e.g. manually setting the system time instead of waiting for actual time to pass.
-    private final long SCHEDULER_TASK_TRIGGER_DELAY_MS = 10;
+    private final long SCHEDULER_TASK_TRIGGER_DELAY_MS = 200;
```
Comment (thilo-behnke) [inline src/test/java/de/tum/in/www1/artemis/service/ProgrammingExerciseScheduleServiceTest.java]:
```suggestion
    private final long SCHEDULER_TASK_TRIGGER_DELAY_MS = 100;
```

**Pull request merged**